pkg/lib/saml/service.go

```diff
@@ -912,3 +964,55 @@ var _ dsig.X509KeyStore = &x509KeyStore{}
 func (x *x509KeyStore) GetKeyPair() (privateKey *rsa.PrivateKey, cert []byte, err error) {
 	return x.privateKey, x.cert, nil
 }
+
+func formatAttribute(raw interface{}) ([]string, error) {
+	// Serialize and Parse it once to ensure it only include types in encoding/json
+	rawBytes, err := json.Marshal(raw)
+	if err != nil {
+		// This should not fail, panic if failed
+		panic(err)
+	}
+	var parsedRaw interface{}
+	err = json.Unmarshal(rawBytes, &parsedRaw)
+	if err != nil {
+		// This should not fail, panic if failed
+		panic(err)
+	}
+	raw = parsedRaw
+
+	formatSlice := func(raw []any) ([]string, error) {
+		values := []string{}
+		for _, itemRaw := range raw {
+			itemValues, err := formatAttribute(itemRaw)
+			if err != nil {
+				return nil, err
+			}
+			values = append(values, itemValues...)
+		}
+		return values, nil
+	}
+
+	formatMap := func(raw map[string]any) ([]string, error) {
+		// If it is an object, output the json stringified object
+		b, err := json.Marshal(raw)
+		if err != nil {
+			return nil, err
+		}
+		return []string{string(b)}, nil
+	}
+
+	switch raw := raw.(type) {
+	case string:
+		return []string{raw}, nil
+	case float64:
+		return []string{fmt.Sprintf("%v", raw)}, nil
+	case bool:
+		return []string{fmt.Sprintf("%v", raw)}, nil
+	case []any:
+		return formatSlice(raw)
+	case map[string]any:
+		return formatMap(raw)
```

Comment:
I think we only need to support primitive types for now.

If the resolved attribute is not a primitive Go data type, then we should fail with an error.

There is a caveat. A JSON number in Go is always deserialized as float64, but an attribute could be an integer. If we just use the deserialized type float64, the output SAML attribute is inaccurate.

Reply:
For primitive types, do you mean string, float64, bool?

I think slice should also be supported, attribute values in xml can always be multiple.

So maybe I will remove the support of map.

Reply:
> There is a caveat. A JSON number in Go is always deserialized as float64, but an attribute could be an integer. If we just use the deserialized type float64, the output SAML attribute is inaccurate.

I think we will simply return [decimal](https://www.w3.org/TR/xmlschema-2/#decimal) for any float64.